tests/test_url.py

```diff
@@ -1104,6 +1104,35 @@ def test_scheme_case_insensitive(self):
         result = parse_data_uri("DaTa:,A%20brief%20note")
         self.assertEqual(result.data, b"A brief note")
 
+    def test_safe_url_string_encode_idna_domain_with_port(self):
+        self.assertEqual(
+            safe_url_string("http://新华网.中国:80"), "http://xn--xkrr14bows.xn--fiqs8s:80"
+        )
+
+    def test_safe_url_string_encode_idna_domain_with_username_password_and_port_number(
+        self,
+    ):
+        self.assertEqual(
+            safe_url_string("ftp://admin:admin@新华网.中国:21"),
+            "ftp://admin:admin@xn--xkrr14bows.xn--fiqs8s:21",
+        )
+        self.assertEqual(
+            safe_url_string("http://Åsa:abc123@➡.ws:81/admin"),
+            "http://%C3%85sa:abc123@xn--hgi.ws:81/admin",
+        )
+        self.assertEqual(
+            safe_url_string("http://japão:não@️i❤️.ws:8000/"),
+            "http://jap%C3%A3o:n%C3%A3o@xn--i-7iq.ws:8000/",
+        )
+
+    def test_safe_url_string_encode_idna_domain_with_username_and_empty_password_and_port_number(
+        self,
+    ):
+        self.assertEqual(
+            safe_url_string("ftp://admin:@新华网.中国:21"),
+            "ftp://admin@xn--xkrr14bows.xn--fiqs8s:21",
+        )
```

Comment:
Can you also add a test for an empty password?

I suspect it will fail, because in that case I think `part.password` is `None`, but your current code seems to ignore the user if `part.password` is not a string.

Reply:
This is already the test for empty password, i'm missing something?

Reply:
Done.